typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,726 @@
+  return { type: "And", l, r, proof: identity };
+};
+
+const interpret: <A>(expr: ArithExpr<A>) => A = (expr) => {
+  switch (expr.type) {
+    case "Num":
+      return expr.proof(expr.v);
+    case "Plus":
+      return expr.proof(interpret(expr.l) + interpret(expr.r));
+    case "Gt":
+      return expr.proof(interpret(expr.l) > interpret(expr.r));
+    case "And":
+      return expr.proof(interpret(expr.l) && interpret(expr.r));
+    default:
+      return expr;
+  }
+};
+
+const testExpr = and(gt(plus(num(23), num(12)), num(170)), gt(num(35), num(47)));
+interpret(testExpr); // false
+
+const wrongExpr = and(num(23), num(12)); // error
+```
+
+Another way to implement eDSL is tagless final.
+Moving from data type to type class we are able to implement the same logic.
+We will use `fp-ts` to implement it.
+
+```typescript
+import { Kind, URIS } from "fp-ts/HKT";
+
+type ArithExpr<Expr extends URIS> = {
+  num: (v: number) => Kind<Expr, number>;
+  plus: (l: Kind<Expr, number>, r: Kind<Expr, number>) => Kind<Expr, number>;
+  gt: (l: Kind<Expr, number>, r: Kind<Expr, number>) => Kind<Expr, boolean>;
+  and: (l: Kind<Expr, boolean>, r: Kind<Expr, boolean>) => Kind<Expr, boolean>;
+};
+
+type Interpreter<A> = {
+  interpret: A;
+};
+
+type ToS<A> = {
+  toString: string;
+};
+
+declare module "fp-ts/lib/HKT" {
+  interface URItoKind<A> {
+    readonly Interpreter: Interpreter<A>;
+    readonly ToS: ToS<A>;
+  }
+}
+
+const arithInterpreter: ArithExpr<"Interpreter"> = {
+  num: (v: number) => {
+    return { interpret: v };
+  },
+  plus: (l: Interpreter<number>, r: Interpreter<number>) => {
+    return { interpret: l.interpret + r.interpret };
+  },
+  gt: (l: Interpreter<number>, r: Interpreter<number>) => {
+    return { interpret: l.interpret > r.interpret };
+  },
+  and: (l: Interpreter<boolean>, r: Interpreter<boolean>) => {
+    return { interpret: l.interpret && r.interpret };
+  },
+};
+
+const arithToS: ArithExpr<"ToS"> = {
+  num: (v: number) => {
+    return { toString: v.toString() };
+  },
+  plus: (l: ToS<number>, r: ToS<number>) => {
+    return { toString: `(${l.toString} + ${r.toString})` };
+  },
+  gt: (l: ToS<number>, r: ToS<number>) => {
+    return { toString: `(${l.toString} > ${r.toString})` };
+  },
+  and: (l: ToS<boolean>, r: ToS<boolean>) => {
+    return { toString: `(${l.toString} && ${r.toString})` };
+  },
+};
+
+const testExpr: <Expr extends URIS>(E: ArithExpr<Expr>) => Kind<Expr, boolean> = (E) =>
+  E.and(E.gt(E.plus(E.num(23), E.num(12)), E.num(170)), E.gt(E.num(35), E.num(47)));
+
+testExpr(arithInterpreter).interpret; // false
+testExpr(arithToS).toString; // (((23 + 12) > 170) && (35 > 47))
+```
+
+As in the previous parts you can also see more complex examples in Yuriy Bogomolov [edsl workshop](https://github.com/YBogomolov/workshop-edsl-in-typescript/tree/master) and Giulio Canti [example](https://gist.github.com/gcanti/9a0c2a666621f03b80457831ff3ab997).
+
+## Conclusion
```

Comment:
Perhaps add a summary of the article here

Reply:
Hm, what kind of summary it should be?

Reply:
Rather generic: what you talked about in the article, some main points repeated, perhaps some links to other places.